exec/java-exec/src/main/java/org/apache/drill/exec/work/batch/IncomingBuffers.java

```diff
@@ -103,8 +104,12 @@ public boolean batchArrived(final IncomingDataBatch incomingBatch) throws Fragme
             Arrays.toString(collectorMap.values().toArray())));
       }
 
+      // Use the Data Collector's buffer allocator if set, otherwise the fragment's one
+      BufferAllocator ownerAllocator =
+        collector.getOprAllocator() != null ? collector.getOprAllocator() : context.getAllocator();
```

Comment:
Consider moving this logic to `AbstractDataCollector.getAllocator()` or `AbstractDataCollector.setAllocator()`.